Add skeleton for ucs-exporter

mithunmgore opened the pull request:
Sample exporter is created

Signed-off-by: Mithun Gore <mithungore@gmail.com>

mithunmgore committed:
Add skeleton for ucs-exporter

Sample exporter is created

Signed-off-by: Mithun Gore <mithungore@gmail.com>

viennaa commented on inventory:
```diff
@@ -0,0 +1,5 @@
+{
+    "servers" : [
+     ]
```
please add a dummy server like:
`server123`

mithunmgore replied:
Done

viennaa commented on modules/main_collector.py:
```diff
@@ -0,0 +1,52 @@
+        self.creds = creds
+        self.handles = dict()
+
+    def __del__(self):
```
you are not actually doing something in here, remove pls

mithunmgore replied:
Done

viennaa commented on modules/main_collector.py:
```diff
@@ -0,0 +1,52 @@
+
+    def collect(self):
+        print("Create handles for UcsmServers. ")
+        info_metric = Metric('ucs_exporter_info', 'information metric, internal exporter info', "gauge")
```
please use GaugeMetricFamily and rename info_metric, usually one is using `g` for Gauges, if you have multiple, then a little bit more.

mithunmgore replied:
Done

viennaa commented on modules/main_collector.py:
```diff
@@ -0,0 +1,52 @@
+
+class MainCollector(object):
+    def __init__(self, servers, creds):
+        self.servers = servers
```
This is too static. If it changes during execution it won't update having it statically defined in here. Have a function which is reading the inventory file every time a collect() is triggered and iterate over the servers returned.

mithunmgore replied:
Added dynamic execution

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,72 @@
+    (options, args) = parser.parse_args()
+    print(options)
+    print(args)
+    return options
```
catch needed options. If user, inventory, config and master_password is not given, the program shouldn't start.

mithunmgore replied:
Added catch.

viennaa commented on modules/main_collector.py:
```diff
@@ -0,0 +1,52 @@
+        # for x in range(self.handles):
+        #     if len(self.handles)>x:
+        #         self.handles[x].logout()
+
```
Add a describe() function to avoid prometheus running collect() initially when the the Collector is registered. See our other exporters.

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,72 @@
+    return options
+
+
+def get_inventory(params):
```
this is the part, which needs to be triggered in collect(). Add a helper class and have this called without instantiating.
Usually I would add this in a Base Class, which your MainCollector is inheriting from. Then it would just be a `server_list = self.server_list`.

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,72 @@
+        creds = {"username": params.user, "password": params.master_password}
+
+        main_collector = MainCollector(servers=server_list, creds=creds)
+    except Exception as e:
```
this try catch is too generic. Catch the actual params where you are expecting them. I had another comment in parse_options about this.

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,72 @@
+    try:
+        REGISTRY.register(main_collector)
+    except Exception as e:
+        print("failed to register: {}".format(params))
```
better error message would be: `failed to register main_collector`

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,72 @@
+
+if __name__ == '__main__':
+    params = get_params()
+    srv = start_http_server(1234)
```
please use prometheus typical port > 9000

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,72 @@
+    srv = start_http_server(1234)
+    start_collector(params)
+    while True:
+        time.sleep(3600)
```
put it to something between 10-60

viennaa commented on prometheus.yml:
```diff
@@ -0,0 +1,9 @@
+global:
```
what is this file being used for? Looks like something we are configuring in our helm-charts later.

viennaa commented on modules/main_collector.py:
```diff
@@ -0,0 +1,52 @@
+        for x in range(len(self.servers)):
+            try:
+                print("Check if handle exists, else login ")
+                if self.handles.get(self.servers[x]):
```
how are you renewing the handles in case they have expired?

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,72 @@
+    try:
+        server_list = get_inventory(params)
+        print(server_list, "List of servers")
+        creds = {"username": params.user, "password": params.master_password}
```
second param should be `master_password` instead of `password`

viennaa commented on modules/ucsm.py:
```diff
@@ -0,0 +1,34 @@
+        self.password = self._get_password()
+        self.handle = self._login()
+
+    def _get_password(self):
```
this could be something generic which is going to be needed elsewhere too. Either put it in a helper class or in the BaseClass, if you decide to go for it. I then would give `UcsmServer` the decrypted pw straight.

viennaa commented on modules/ucsm.py:
```diff
@@ -0,0 +1,34 @@
+        try:
+            handle.login()
+        except UcsException as e:
+            print(e)
```
`print('Error logging in', self.ucsm_server, str(e))`
give it a `return False`

viennaa commented on modules/main_collector.py:
```diff
@@ -0,0 +1,52 @@
+                    continue
+                else:
+                    srv_obj = UcsmServer(self.servers[x], self.creds['username'], self.creds['password'])
+                    self.handles[self.servers[x]] = srv_obj.handle
```
before adding, check for the `False return` and don't add it.

viennaa commented on modules/main_collector.py:
```diff
@@ -0,0 +1,52 @@
+
+
+        for x in range(len(self.servers)):
+            try:
```
remove this way to generic try catch.

viennaa commented on modules/main_collector.py:
```diff
@@ -0,0 +1,52 @@
+                print(traceback.format_exc())
+
+                for handle in self.handles:
+                    del handle
```
can't the handles be reused?

mithunmgore replied:
Reusing the handles.

viennaa commented on modules/main_collector.py:
```diff
@@ -0,0 +1,52 @@
+                    }, value=1)
+        yield info_metric
+
+        ucs_license_collector = UcsServerLicense(self.handles)
```
You are naming it as a collector, but not handling it as a dedicated collector. This needs to go to a dedicated collector. Please rename: `MainCollector` to better fit what it is really doing and add a second collector `UcsLicenseCollector` which takes over from here on.

mithunmgore replied:
Renamed it to ucsm collector, considering there will be another ucsc collector.

viennaa commented on modules/server_license.py:
```diff
@@ -0,0 +1,20 @@
+from prometheus_client import Metric
+
+
+class UcsServerLicense:
```
This is basically the UcsServerLicenseCollector collector already. Please also treat it accordingly.

mithunmgore replied:
Not sure exactly, what you want to say. I renamed it as collector

viennaa commented on modules/server_license.py:
```diff
@@ -0,0 +1,20 @@
+
+    def collect(self):
+        print("Get license info of ports ")
+        license_state_metric = Metric('ucs_license', 'Information about port license', "gauge")
```
use GaugeMetricFamily

mithunmgore replied:
Yes changed from Gauge to Gaugemetricfamily everywhere

viennaa commented on modules/server_license.py:
```diff
@@ -0,0 +1,20 @@
+    def collect(self):
+        print("Get license info of ports ")
+        license_state_metric = Metric('ucs_license', 'Information about port license', "gauge")
+        for server, handle in self.handles.items():
```
and here it comes in handy, if you have the handles stored in a `BaseCollector`

mithunmgore replied:
Added handles  in UcsmCollector.

viennaa reviewed (changes requested):
Thanks for pushing it so far, the general approach is good. My annotations are inline. 

Please really consider adding a BaseCollector which hosts things like `handle` handling (initial login, refreshing etc) and `master_password`. There are some comments in this direction.

mithunmgore committed:
Address the comments on PR

Addressed comments.

Signed-off-by: Mithun Gore <mithungore@gmail.com>

mithunmgore commented:
@viennaa I have addressed almost all the comments, Please take a look and let me know If I have missed or understood it correctly.
Thanks for all the comments and depth review.

viennaa commented on collectors/UcsmCollector.py:
```diff
@@ -0,0 +1,61 @@
+    def describe(self):
+        yield GaugeMetricFamily("ucsm_metrics", "ucsm_collector_registered")
+
+    def get_inventory(self):
```
This needs to go to `BaseCollector` class

viennaa commented on collectors/ServerLicenseCollector.py:
```diff
@@ -0,0 +1,26 @@
+from prometheus_client.core import GaugeMetricFamily
+
+
+class UcsServerLicenseCollector():
```
This should be inheriting from `BaseCollector`

viennaa commented on collectors/UcsmCollector.py:
```diff
@@ -0,0 +1,61 @@
+        g.add_metric(labels=["ucs_get_handles"], value=1)
+        yield g
+
+        ucs_license_collector = UcsServerLicenseCollector(self.handles)
```
Remove these lines, we're going to register this collector in our main routine straight away.

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,58 @@
+
+if __name__ == '__main__':
+    params = get_params()
+    srv = start_http_server(9876)
```
swap it with L 56, register the collectors before launching the http server.

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,58 @@
+    creds = {"username": params['user'], "master_password": params['master_password']}
+    inventory_file = params["inventory"]
+    ucsm_collector = UcsmCollector(creds=creds, inventory_file=inventory_file)
+    REGISTRY.register(ucsm_collector)
```
Here you need to register your second collector too (UcsServerLicenseCollector). This way you are able to scale up horizontally if needed and add as many collectors as necessary.

viennaa commented on collectors/ServerLicenseCollector.py:
```diff
@@ -0,0 +1,26 @@
+            mo = handle.query_dn(dn)
+
+            license_state = mo.lic_state
+            labels = [server, "27"]
```
What is `27`? I get the feeling, this needs to be configurable instead of hard coded.

viennaa commented on collectors/UcsmCollector.py:
```diff
@@ -0,0 +1,61 @@
+            except UcsException as e:
+                print(e, ", May be check for username and password !")
+                return
+        g.add_metric(labels=["ucs_get_handles"], value=1)
```
You are only adding value `1`. I'd expect having value 0 added, if it is not as expected or not reachable. Configure a default value before your loop which is 0, and only set it to 1, when you are hitting, what you are looking fore. Then, at the end, you set it to the variable (which either is 1 or 0 then).

viennaa replied:
Or am I mistaking this metric? The label is pretty generic, usually you would add `server_name` or `ucs_cluster_name` or something what enables you to see, where this metric is coming from to properly alert.

viennaa commented on collectors/UcsmCollector.py:
```diff
@@ -0,0 +1,61 @@
+        g = GaugeMetricFamily('ucsm_exporter_info', 'information metric, internal exporter info')
+
+        print("Create handles for Ucsm Servers.")
+        for x in range(len(server_list)):
```
rename x to what it actually is

viennaa commented on collectors/ServerLicenseCollector.py:
```diff
@@ -0,0 +1,26 @@
+from prometheus_client.core import GaugeMetricFamily
```
Rename the file please to macht the class name

viennaa commented on collectors/UcsmCollector.py:
```diff
@@ -0,0 +1,61 @@
+        g = GaugeMetricFamily('ucsm_exporter_info', 'information metric, internal exporter info')
+
+        print("Create handles for Ucsm Servers.")
+        for x in range(len(server_list)):
```
The handle handling part should actually go to your BaseClass. There you do the login. Always call a BaseClass variable which is holding the handle. Could easily be your dict. Therefore you would be able to use it more than once. 
If the handle is not working anymore, you need to reinitiate the login procedure. For now the easiest way is to implement a retry function in case the handle is not working properly. (here: `LicenceCollectore  mo = handle.query_dn(dn)`

Catch the error which occurs with an expired handle and then have a the login function called again to overwrite the old handle.

A better implementation however would be, to provide some sort of an interface. We take care of this after you've implemented this.

mithunmgore replied:
@viennaa what I am thinking is
1. Handle handling in BaseClass: The handle for ucsm and ucsc differes like, ucsm needs UcsHandle handle from ucsmsdk, and ucsc needs UcscHandle from ucscsdk. then for login, handle.login doesnt accept the timeout for ucsc. So considering this I was thinking to have these handles in UcsmCollector and UcscCollector. let me know, what you think?

If you think to have it in Baseclass then we need to handle that in login function.

2. Regarding handle reinitiation, If you check in L41-L48 I am checking for the handle if exists, then again I am doing login on the same. I checked this works even if the VPN breaks in-between and then we can go ahead and collect the metrics. I will surely handle the exception at query_dn. Let me know if think this is what you want.

viennaa replied:
1. I am just seeing `UcsHandle` being used: https://github.com/sapcc/ucs-exporter/pull/1/files#diff-6d0b49c065163976a8157419af4bbc03R18
no sign of the UcsmHandle you are referring to.
I just designed a bigger draft, how this could work with some more classes, but I get the feeling it is maybe a little bit too much, so we go with the handle approach like you have it.

2. not really, but it is hard to implement with the current structure I guess. But now that I am looking at it again, you are triggering a login at each collect() run again (L42), since the condition always matches after it has been initialised initially. Ok guess that should work without a retry. 

3. what about closing the connection at the end, shouldn't there be some logout? I guess the handle timeout is something like 1h or so, so it might lock up further loggins, if we have to many (lose) open handles.

viennaa commented on ucs-exporter.py:
```diff
@@ -0,0 +1,58 @@
+
+    (options, args) = parser.parse_args()
+    options = vars(options)
+    for option in options:
```
Smart move, I like it :)

mithunmgore replied:
Thanks :)

viennaa commented on modules/ucsm.py:
```diff
@@ -0,0 +1,25 @@
+
+class UcsmServer(BaseClass):
+    def __init__(self, ucs_server, user, master_password):
+        super().__init__(ucs_server, user, master_password)
```
Like this idea, nice!

mithunmgore replied:
Thanks 👍

viennaa commented on modules/ucsm.py:
```diff
@@ -0,0 +1,25 @@
+from BaseModule import BaseClass
+
+
+class UcsmServer(BaseClass):
```
Please rename file to match class name

viennaa reviewed (changes requested):
You've brought in some cool ideas. Some other things popped up now, but it's going in the right direction.

mithunmgore committed:
Address PR comments

Addressed comments and modfified code, renamed files

Signed-off-by: Mithun Gore <mithungore@gmail.com>

viennaa commented on BaseCollector.py:
```diff
@@ -1,11 +1,64 @@
+        server_list = self.get_inventory()
+        print("Create handles for Ucsm Servers.")
+        for server in range(len(server_list)):
+            try:
```
remove this try-except block. It Would only be catching a login-error when running UcsmServer. In there you have try-excepted the login method already, so no need to do it again.

viennaa commented on ucs-exporter.py:
```diff
@@ -46,13 +52,16 @@ def start_collector(params):
+    # Register collectors
+    for collector, class_name in COLLECTORS.items():
+        REGISTRY.register(getattr(import_module("collectors.{}".format(
+                                collector)), class_name)(creds, inventory_file))
```
You don't need to use a dict here. Your dict looks like:

```
{ "a": "a", "b" : "b"}
```

and this should be a list [ "a", "b"]. Then loop through it.

mithunmgore replied:
This is done.

viennaa commented on ucs-exporter.py:
```diff
@@ -46,13 +52,16 @@ def start_collector(params):
 if __name__ == '__main__':
     params = get_params()
-    srv = start_http_server(9876)
     start_collector(params)
```
maybe we rename this now to `register_collectors`

mithunmgore replied:
Done

mithunmgore committed:
Rename function and change dict to list

Renamed function to match the cause and removed dict and added list of collectors

Signed-off-by: Mithun Gore <mithungore@gmail.com>

mithunmgore committed:
Move exception from basecollector to ucsmserver

Moved exceptions

Signed-off-by: Mithun Gore <mithungore@gmail.com>

viennaa commented on modules/UcsmServer.py:
```diff
@@ -19,7 +19,11 @@ def _login(self):
+            return
         except UcsException as e:
-            raise e
+            print(e, ", May be check for username and password !")
```
Please change these error messages to:
`print("Problem logging in to", self.ucs_server, ":", e)`
Something similar with OSError too.

mithunmgore replied:
Changed. kept the message same for both errors and "e" will describe it, like "Authentication Error " and "Url Error"

viennaa replied:
yep, fine with me.

mithunmgore committed:
Change exception message

Added meaningfull message

Signed-off-by: Mithun Gore <mithungore@gmail.com>

viennaa commented on modules/UcsmServer.py:
```diff
@@ -20,10 +20,10 @@ def _login(self):
             handle.login(timeout=5)
         except OSError as e:
-            print(e, "May be server not reachable !")
+            print("Problem logging in to", self.ucs_server, ":", e)
```
Forgot to mention, you have to use `str(e)` instead of `e`

mithunmgore committed:
typecast error to string

Done

Signed-off-by: Mithun Gore <mithungore@gmail.com>

viennaa reviewed (approved):
Thanks for your commitment, good to from here on.
